### PR TITLE
feat(subscription): support blocking cursor

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

#18208 

In this pr, we support blocking subscription cursor and all cursor timeout.
the stmt of our fetch cursor becomes 
`fetch next/num from cursor_name with (timeout = 'xx');`

about subscription cursor's `timeout `:
1. When there are a lot of values in the subscription and the fetch is set to a large number of rows, the fetch statement may take longer than `timeout` time to execute. This time we will returns all the values currently fetched.
2. When subscription has no new data at the moment and fetch is used, there are two scenarios

-  If you have already fetched more than one row, the fetch statement does not block, but returns all the values currently fetched.
- If there is currently no value, we block until there is a new value, or until the blocking time exceeds `timeout`

about query cursor's `timeout `:
only support 1

When `with (timeout = 'xx')` is not set, 
The query timeout is u64::max
The block timeout is 0

`timeout` is a string in interval format.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Support 
`fetch next/num from cursor_name with (timeout = 'xx');`
About subscription cursor's `timeout `:
1. When there are a lot of values in the subscription and the fetch is set to a large number of rows, the fetch statement may take longer than `timeout` time to execute. This time we will returns all the values currently fetched.
2. When subscription has no new data at the moment and fetch is used, there are two scenarios
-  If we have already fetched more than one row, the fetch statement does not block, but returns all the values currently fetched.
- If there is currently no value, we block until there is a new value, or until the blocking time exceeds `timeout`

About query cursor's `timeout `:
only support 1

When `with (timeout = 'xx')` is not set, 
In the case of 1, this corresponds to timeout = u64::max 
In the case of 2, this corresponds to timeout = 0

`timeout` is a string in interval format.

### CODE REVIEW

File: src/frontend/src/observer/observer_manager.rs [hunk 1]
```diff
@@ -468,6 +512,30 @@ impl FrontendObserverNode {
         self.hummock_snapshot_manager.update(deltas);
     }
 
+    fn handle_cursor_notify(&self, table_ids: Vec<TableId>) {
+        if table_ids.is_empty() {
+            return;
+        }
+        for session in self.sessions_map.read().values() {
```
Comment (hzxa21) [inline src/frontend/src/observer/observer_manager.rs:519]: Currently one CursorNotifier per session and within each CursorNotifier, we have one channel per table. Assume we have N sessions and M tables in total, we can end up creating N*M channels. Not to mention that we have a task in CursorNotifier to maintain the per table channel. These seem inefficient and complicated. IMO, a better way to implement the notification is:
- Create one and only one watch channel to listen on table changed log updates.
- When cursor needs to wait for changelog epoch changes for a table, it only needs to subscribe to this channel.

Implementation-wise, maybe we can store HummockSnapshotManager in each cursor instance and implement a wait_for_epoch similar to the [method](https://github.com/risingwavelabs/risingwave/blob/c47a2e8fce4a1a5bbd4f422b816be2b18460215f/src/frontend/src/scheduler/snapshot.rs#L257) here.

File: src/frontend/src/session/cursor_manager.rs [hunk 2]
```diff
@@ -471,16 +487,31 @@ impl SubscriptionCursor {
                     ans.push(row);
                 }
                 None => {
```
Comment (hzxa21) [inline src/frontend/src/session/cursor_manager.rs:489]: Does `None` here always mean no data? I saw that we will return None when field mismatches [here](https://github.com/risingwavelabs/risingwave/blob/c47a2e8fce4a1a5bbd4f422b816be2b18460215f/src/frontend/src/session/cursor_manager.rs#L368).

File: src/storage/hummock_sdk/src/frontend_version.rs [hunk 3]
```diff
@@ -161,8 +162,8 @@ impl FrontendHummockVersionDelta {
                             truncate_epoch: change_log_delta.truncate_epoch,
                             new_log: change_log_delta.new_log.as_ref().map(|new_log| {
                                 EpochNewChangeLogCommon {
-                                    new_value: vec![],
-                                    old_value: vec![],
+                                    new_value: new_log.new_value.iter().map(|_| ()).collect(),
```
Comment (wenym1) [inline src/storage/hummock_sdk/src/frontend_version.rs:165]: `vec![(); new_log.new_value.len()]`

And the `old_value` incorrectly use the `new_log.new_value`.

File: src/storage/hummock_sdk/src/frontend_version.rs [hunk 4]
```diff
@@ -194,8 +195,16 @@ impl FrontendHummockVersionDelta {
                         table_id.table_id,
                         PbChangeLogDelta {
                             new_log: delta.new_log.as_ref().map(|new_log| PbEpochNewChangeLog {
-                                old_value: vec![],
-                                new_value: vec![],
+                                old_value: new_log
+                                    .old_value
+                                    .iter()
+                                    .map(|_| mock_pb_sstable_info())
```
Comment (wenym1) [inline src/storage/hummock_sdk/src/frontend_version.rs:201]: Can use `vec![PbSstableInfo::default(); new_log.old_value.len()]`.

File: src/storage/hummock_sdk/src/change_log.rs [hunk 5]
```diff
@@ -101,7 +101,7 @@ impl<T> TableChangeLogCommon<T> {
         &self.0[start..end]
     }
 
-    /// Returns epochs where value is non-null and >= `min_epoch`.
+    // Returns epochs where value is non-null and >= `min_epoch`.
```
Comment (wenym1) [inline src/storage/hummock_sdk/src/change_log.rs:104]: Any reason to turn `///` to `//`?

File: src/frontend/src/observer/observer_manager.rs (hunk 1 above)
Comment (wenym1) [reply src/frontend/src/observer/observer_manager.rs:519]: Actually, I think the current cursor implementation in frontend can be greatly simplified. The most complicated part in the current cursor is the state machine. The async feature of rust is a very powerful tool to generate state machine. So for a cursor, it can be implemented as a stream
```
#[try_stream(...)]
async fn cursor_stream(...) {
    if need_snapshot {
         for row in create_snapshot_iter(table_id) { yield row; }
    }
    let mut current_epoch = ...;
    loop {
        for row in change_log(current_epoch, table_id) { yield row; }
        current_epoch = snapshot_manager.wait_next_epoch(current_epoch, table_id);
    }
}
```

When a cursor is created, we can create such stream and hold it in the session. On a `fetch`, we just call `cursor_stream.next().await`, and the timeout can be applied with simply `timeout(timeout_duration, cursor_stream.next()).await`.

We can be aware of table drop in `snapshot_manager`, when we see that the `table_id` does not exist in the `state_table_info` of `FrontendHummockVersion` anymore, and then changes on catalog in this PR can be totally avoided.

File: src/frontend/src/session/cursor_manager.rs (hunk 2 above)
Comment (xxhZs) [reply src/frontend/src/session/cursor_manager.rs:489]: Yes, none means that there is no data for this call to next_row

File: src/frontend/src/session/cursor_manager.rs [hunk 6]
```diff
@@ -471,16 +483,36 @@ impl SubscriptionCursor {
...
-                    break;
+                    if cur == 0 {
+                        // It's only blocked when there's no data
+                        // This method will only be called once, either to trigger a timeout or to get the return value in the next loop via `next_row`.
+                        session
+                            .env
+                            .hummock_snapshot_manager()
+                            .wait_table_change_log_notification(
```
Comment (hzxa21) [inline src/frontend/src/session/cursor_manager.rs:492]: When `wait_table_change_log_notification` timeout triggers, we are relying on the if condition in L504 to hit to exit the while loop. This is somehow strange and dangerous. I susggest we make the following changes:
1. Make `wait_table_change_log_notification` to wait for notification without a timeout.
2. Call `tokio::time::timeout(timeout_seconds, wait_table_change_log_notification)` here.
3. When 2 returns `time::error::Elapsed`, break the loop directly.

File: src/frontend/src/session/cursor_manager.rs [hunk 7]
```diff
@@ -471,16 +483,36 @@ impl SubscriptionCursor {
                     ans.push(row);
                 }
                 None => {
-                    break;
+                    if cur == 0 {
```
Comment (hzxa21) [inline src/frontend/src/session/cursor_manager.rs:486]: We can check whether timeout is > 0 or unset here as well to early return.

File: e2e_test/subscription/main.py [hunk 8]
```diff
@@ -56,7 +57,7 @@ def test_cursor_snapshot():
     execute_insert("declare cur subscription cursor for sub full",conn)
     row = execute_query("fetch next from cur",conn)
     check_rows_data([1,2],row[0],"Insert")
-    row = execute_query("fetch next from cur",conn)
+    row = execute_query("fetch next from cur with (timeout = '2s')",conn)
```
Comment (hzxa21) [inline e2e_test/subscription/main.py:60]: After making timeout = 0 the default, we can revert the changes here.

**Review by hzxa21 — APPROVED** [review]
LGTM
